### PR TITLE
Real time doc capture on mobiles

### PR DESCRIPTION
**Pull request by kopijunkie**
# Task
Implement real-time doc capture for web SDK on mobile following on from PoC in CX-3759.

# Acceptance Criteria: 
* `forceCrossDevice` should work the same
* live capture will show up on mobile devices (default for desktops is document uploader)
* fallback should be the same as selfie on mobile, consider the following cases and match with selfie:
    * camera stop working
    * there is no camera
    * long period of inactivity
    * there is no camera api
* camera permission with same behaviour as selfie
* user tracking similar to selfie
* Passport will have ID3 frame size
* DL and national ID will have ID1 frame size

## Checklist
_put `n/a` if item is not relevant to PR changes_

- [x] Have the CHANGELOG, README, MIGRATION, RELEASE_GUIDELINES docs been updated?
- [ ] Have new automated tests been implemented?
- [ ] Have new manual tests been written down?
- [n/a] Have any new strings been translated?

### CODE REVIEW

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/Photo/DocumentLiveCapture.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+              disabled={hasCameraError}
+              onClick={this.captureDocumentPhoto}
+              className={style.btn}
+            />
+          </div>
+        </Camera>}
+      </div>
+    )
+  }
+}
```
Comment (josokinas) [inline src/components/Photo/DocumentLiveCapture.js]: [Minor] This here indicates a missing newline at the end of the file.

File: src/components/Photo/DocumentLiveCapture.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    const documentCapture = {
+      blob,
+      sdkMetadata,
+      filename: `document_capture.${mimeType(blob)}`
+    }
+    this.props.onCapture(documentCapture)
+  }
+
+  captureDocumentPhoto = () => {
+    this.setState({ isLoading: true })
```
Comment (josokinas) [inline src/components/Photo/DocumentLiveCapture.js]: Don't you have to unset `isLoading` in case of an error?

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/Photo/DocumentLiveCapture.js (hunk 2 above)
Comment (kopijunkie) [reply src/components/Photo/DocumentLiveCapture.js]: There is actually no service call happening here, the loading state is just to compensate for a slight delay of about 1 second when transitioning between the capture screen and the "confirm image" screen on lower end mobiles. 

But I'll add code to unset `isLoading` on `componentWillUnmount` anyway.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://802-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/Photo/DocumentLiveCapture.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+export default class DocumentLiveCapture extends Component<Props, State> {
+  webcam = null
+
+  state: State = {
+    hasBecomeInactive: false,
+    hasCameraError: false,
+    isLoading: false
+  }
+
+  handleTimeout = () => this.setState({ hasBecomeInactive: true })
```
Comment (stefaniacardenas) [inline src/components/Photo/DocumentLiveCapture.js]: There is a lot of code shared here shared between this component and the selfie photo one. For example, this camera errors are very similar and the capture part as well. Maybe we can extract it into a component

File: src/components/utils/camera.js [hunk 4]
```diff
@@ -10,6 +10,20 @@ export const screenshot = (webcam, callback) => {
   canvasToBlob(canvas, blob => callback(blob, sdkMetadata))
 }
 
+export const takePhoto = (webcam, callback) => {
+  const sdkMetadata = getDeviceInfo(webcam.stream)
+  const videoTrack = webcam.stream.getVideoTracks()[0] || {}
+  const imageCapture = new ImageCapture(videoTrack)
```
Comment (stefaniacardenas) [inline src/components/utils/camera.js]: Nice, hopefully there will be more support for this

File: src/components/Capture/Document.js [hunk 5]
```diff
@@ -59,22 +61,39 @@ class Document extends Component {
...
+      return (
+        <DocumentAutoCapture
+          {...propsWithErrorHandling}
+          renderTitle={ renderTitle }
+          renderFallback={ renderFallback }
+          containerClassName={ style.documentContainer }
+          onValidCapture={ this.handleCapture }
+        />
+      )
+    } else if (useLiveDocumentCapture && hasCamera) {
```
Comment (stefaniacardenas) [inline src/components/Capture/Document.js]: Minor: I personally find it more readable to use `if` when the previous condition has a return

File: src/components/Photo/DocumentLiveCapture.js (hunk 3 above)
Comment (kopijunkie) [reply src/components/Photo/DocumentLiveCapture.js]: Yeah, that's true. I guess it can be done as a separate ticket.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,9 @@ This project adheres to the Node [default version scheme](https://docs.npmjs.com
 
 ## [Next version]
 
+### Added
+- Public: `useLiveDocumentCapture` option added to the document capture step
+
```
Comment (stefaniacardenas) [inline CHANGELOG.md]: Are we not going to document this in the README?

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

File: CHANGELOG.md (hunk 6 above)
Comment (kopijunkie) [reply CHANGELOG.md]: Not for now/this ticket. There is a follow up ticket where I'll discuss with See Wah how we want to document it in the README for beta testing.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-802-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://802-pr-onfido-sdk-ui-onfido.surge.sh

**Pull request merged**